rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -0,0 +1,472 @@
+};
+
+struct node_data {
+    unsigned int len;
+    struct data_elem *ary;
+};
+
+static struct trie_node *pfx_table_get_root(const struct pfx_table *pfx_table, const enum lrtr_ip_version ver);
+static int pfx_table_del_elem(struct node_data *data, const unsigned int index);
+static int pfx_table_create_node(struct trie_node **node, const struct pfx_record *record);
+static int pfx_table_append_elem(struct node_data *data, const struct pfx_record *record);
+static struct data_elem *pfx_table_find_elem(const struct node_data *data, const struct pfx_record *record, unsigned int *index);
+static bool pfx_table_elem_matches(struct node_data *data, const uint32_t asn, const uint8_t prefix_len);
+static void pfx_table_notify_clients(struct pfx_table *pfx_table, const struct pfx_record *record, const bool added);
+static int pfx_table_remove_id(struct pfx_table *pfx_table, struct trie_node **root, struct trie_node *node, const struct rtr_socket *socket, const unsigned int level);
+static int pfx_table_node2pfx_record(struct trie_node *node, struct pfx_record records[], const unsigned int ary_len);
+static void pfx_table_free_reason(struct pfx_record **reason, unsigned int *reason_len);
+
+
+void pfx_table_notify_clients(struct pfx_table *pfx_table, const struct pfx_record *record, const bool added)
+{
+    if(pfx_table->update_fp != NULL)
+        pfx_table->update_fp(pfx_table, *record, added);
+}
+
+void pfx_table_init(struct pfx_table *pfx_table, pfx_update_fp update_fp)
+{
+    pfx_table->ipv4 = NULL;
+    pfx_table->ipv6 = NULL;
+    pfx_table->update_fp = update_fp;
+    pthread_rwlock_init(&(pfx_table->lock), NULL);
+}
+
+void pfx_table_free(struct pfx_table *pfx_table)
+{
+    for(int i = 0; i < 2; i++) {
+        struct trie_node *root = (i == 0 ? pfx_table->ipv4 : pfx_table->ipv6);
+        if(root != NULL) {
+            struct trie_node *rm_node;
+
+            pthread_rwlock_wrlock(&(pfx_table->lock));
+            do {
+                struct node_data *data = (struct node_data *) (root->data);
+                for(unsigned int i = 0; i < data->len; i++) {
+                    struct pfx_record record = { data->ary[i].asn, (root->prefix), root->len, data->ary[i].max_len, data->ary[i].socket};
+                    pfx_table_notify_clients(pfx_table, &record, false);
+                }
+                rm_node = (trie_remove(root, &(root->prefix), root->len, 0));
+                assert(rm_node != NULL);
+                free(((struct node_data *) rm_node->data)->ary);
+                free(rm_node->data);
+                free(rm_node);
+            } while(rm_node != root);
+            if(i == 0)
+                pfx_table->ipv4 = NULL;
+            else
+                pfx_table->ipv6 = NULL;
+
+            pthread_rwlock_unlock(&(pfx_table->lock));
+        }
+    }
+    pthread_rwlock_destroy(&(pfx_table->lock));
+}
+
+int pfx_table_append_elem(struct node_data *data, const struct pfx_record *record)
```

Comment:
Could be possible that you're calling this function only from locked code sections.